aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

```diff
@@ -0,0 +1,243 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::natives::{
+    aggregator_natives::helpers_v2::{
+        aggregator_snapshot_string_info, aggregator_snapshot_u128_info,
+        aggregator_snapshot_u64_info,
+    },
+    AccountAddress,
+};
+use aptos_gas_schedule::gas_params::natives::aptos_framework::*;
+use aptos_native_interface::{
+    safely_pop_arg, RawSafeNative, SafeNativeBuilder, SafeNativeContext, SafeNativeResult,
+};
+use aptos_types::vm_status::StatusCode;
+use move_binary_format::errors::PartialVMError;
+use move_core_types::value::{MoveStructLayout, MoveTypeLayout};
+use move_vm_runtime::native_functions::NativeFunction;
+use move_vm_types::{
+    loaded_data::runtime_types::Type,
+    values::{Struct, StructRef, Value},
+};
+use smallvec::{smallvec, SmallVec};
+use std::{collections::VecDeque, ops::Deref};
+
+/***************************************************************************************************
+ * native fun create_snapshot(value: Element): AggregatorSnapshot<Element>;
+ **************************************************************************************************/
+
+fn native_create_snapshot(
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(ty_args.len(), 1);
+    debug_assert_eq!(args.len(), 1);
+    context.charge(AGGREGATOR_V2_CREATE_SNAPSHOT_BASE)?;
+
+    match ty_args[0] {
+        Type::U128 => {
+            let input = safely_pop_arg!(args, u128);
+            Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u128(
+                input
+            )]))])
+        },
+        Type::U64 => {
+            let input = safely_pop_arg!(args, u64);
+            Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u64(
+                input
+            )]))])
+        },
+        _ => {
+            // Check if the type is a string
+            let ty = context
+                .deref()
+                .type_to_fully_annotated_layout(&ty_args[0])?;
+            if let MoveTypeLayout::Struct(MoveStructLayout::WithTypes { type_, .. }) = ty {
+                if type_.name.as_str() == "String"
+                    && type_.module.as_str() == "string"
+                    && type_.address == AccountAddress::ONE
+                {
+                    let input = safely_pop_arg!(args, Struct)
+                        .unpack()?
+                        .next()
+                        .unwrap()
+                        .value_as::<Vec<u8>>()?;
+                    let move_string_value =
+                        Value::struct_(Struct::pack(vec![Value::vector_u8(input)]));
+                    let move_snapshot_value = Value::struct_(Struct::pack(vec![move_string_value]));
+                    return Ok(smallvec![move_snapshot_value]);
+                }
+            }
+            // If not a string, return an error
+            Err(PartialVMError::new(StatusCode::ABORTED)
```

Comment:
I think for natives one should use `SafeNativeError`, e.g.
```
return Err(SafeNativeError::Abort {
    abort_code: 0x02_0005,
});
```

Reply:
The `.into()` method converts the `PartialVMError` into `SafeNativeError`.

Reply:
I think the into part is there because there were cases where error handling was not done properly. Why not do it with `SafeNativeError` from the start? Maybe @vgao1996 knows more here

Reply:
Done.